### ISSUE TITLE
Generalize append and prepend

### ISSUE THREAD

**Issue by gabejohnson** [sanctuary-js/sanctuary#343]
Currently, we have `append :: a -> Array a -> Array a` and `prepend :: a -> Array a -> Array a`.

There's no reason this can't be generalized to any type that implements Semigroup and Applicative (or really lawful `concat` and `of`).

Then we can apply them to `Immutable.List`:
```js
S.prepend(1, S.append(3, Z.of(List, 2))) // List(1, 2, 3)
```

**Comment by davidchambers** [comment]
Do you fancy the idea of submitting a pull request? :)

**Comment by gabejohnson** [comment]
I can do that, but I'm not quite sure how to type it in the description. `append :: a -> (Semigroup & Applicative) a -> (Semigroup & Applicative) a`. I don't know the notation for "implements n algebras".

**Comment by davidchambers** [comment]
The type signature you're suggesting would look like this, @gabejohnson:

```purescript
append :: (Semigroup f, Applicative f) => a -> f a -> f a
```

With sanctuary-def:

```javascript
S.append =
def('append',
    {f: [Z.Semigroup, Z.Applicative]},
    [a, f(a), f(a)],
    append);
```

> I think this can be Foldable

That would be even nicer. :)

**Comment by gabejohnson** [comment]
> I think this can be Foldable

I'm not sure I'm tracking. @vendethiel what's your suggested implementation?

**Comment by vendethiel** [comment]
I'm wrong. It looks like it needs `Alt` (not Alternative because of the annihilation law), or maybe `Plus` for `empty`.

**Comment by svozza** [comment]
Implemented in #345.
